### PR TITLE
Upgrade AdminUser to APIP 3

### PR DESCRIPTION
**Pull request by TheMilek**
| Q               | A
|-----------------|-----
| Branch?         | api-platform-3 
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| License         | MIT

### CODE REVIEW

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/state_processors.xml [hunk 1]
```diff
@@ -16,6 +16,14 @@
            xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd"
 >
     <services>
+        <service id="Sylius\Bundle\ApiBundle\StateProcessor\AdminUserProcessor">
```
Comment (mpysiak) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services/state_processors.xml]:
```suggestion
        <service id="sylius.api.state_processor.admin_user" class="Sylius\Bundle\ApiBundle\StateProcessor\AdminUserProcessor">
```

File: src/Sylius/Bundle/ApiBundle/StateProcessor/AdminUserProcessor.php [hunk 2]
```diff
@@ -11,43 +11,46 @@
...
 
-    public function persist($data, array $context = [])
-    {
         $this->passwordUpdater->updatePassword($data);
 
-        return $this->decoratedDataPersister->persist($data, $context);
-    }
+        if ($operation instanceof DeleteOperationInterface) {
+            if ($this->isTryingToDeleteLoggedInUser($data)) {
+                throw new CannotRemoveCurrentlyLoggedInUser();
```
Comment (diimpp) [inline src/Sylius/Bundle/ApiBundle/StateProcessor/AdminUserProcessor.php]: Would be cool to sync terminology.

Either
```suggestion
            if ($this->isTryingToRemoveLoggedInUser($data)) {
                throw new CannotRemoveCurrentlyLoggedInUser();
```
or
```suggestion
            if ($this->isTryingToDeleteLoggedInUser($data)) {
                throw new CannotDeleteCurrentlyLoggedInUser();
```

**Review by mpysiak — APPROVED** [review]

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/AdminUser.xml [hunk 3]
```diff
@@ -16,7 +16,7 @@
     xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
     xsi:schemaLocation="https://api-platform.com/schema/metadata/resources-3.0 https://api-platform.com/schema/metadata/resources-3.0.xsd"
 >
-    <resource class="Sylius\Component\Core\Model\AdminUser" shortName="Administrator">
+    <resource class="Sylius\Component\Core\Model\AdminUser" processor="sylius.api.state_processor.admin_user" shortName="Administrator" routePrefix="admin">
```
Comment (Wojdylak) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/AdminUser.xml]: If I understand correctly, this processor is only for the delete action, so maybe we should do it like [this](https://github.com/Sylius/Sylius/pull/16398#discussion_r1639521794)

File: tests/Api/Admin/AdminUsersTest.php [hunk 4]
```diff
@@ -15,11 +15,148 @@
...
+        $this->assertResponse($response, 'admin/admin_user/log_in_admin_response', Response::HTTP_OK);
+    }
+
+    /** @test */
+    public function it_gets_administrators(): void
+    {
+        $this->loadFixturesFromFiles(['authentication/api_administrator.yaml', 'administrator.yaml']);
+        $header = array_merge($this->logInAdminUser('api@example.com'), self::CONTENT_TYPE_HEADER);
+
+        $this->client->request(method: 'GET', uri: '/api/v2/admin/administrators', server: $header);
```
Comment (GSadee) [inline tests/Api/Admin/AdminUsersTest.php]: You can use here `headerBuilder` and `requestGet` method:
```suggestion
        $header = $this->headerBuilder()->withJsonLdAccept()->withAdminUserAuthorization('api@example.com')->build();

        $this->requestGet(uri: '/api/v2/admin/administrators', headers:: $header);
```

File: tests/Api/Admin/AdminUsersTest.php [hunk 5]
```diff
@@ -15,11 +15,148 @@
...
+
+        $this->assertResponse(
+            $this->client->getResponse(),
+            'admin/admin_user/get_administrators_response',
+            Response::HTTP_OK,
+        );
```
Comment (GSadee) [inline tests/Api/Admin/AdminUsersTest.php:58]: Passing `Response::HTTP_OK` is not necessary as it is a default value
```suggestion
        $this->assertResponse($this->client->getResponse(), 'admin/admin_user/get_administrators_response');
```

File: src/Sylius/Bundle/ApiBundle/StateProcessor/Put/AdminUserProcessor.php [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    {
+        Assert::isInstanceOf($data, AdminUserInterface::class);
+
+        if ($operation instanceof DeleteOperationInterface) {
+            return $this->persistProcessor->process($data, $operation, $uriVariables, $context);
+        }
+
+        $this->passwordUpdater->updatePassword($data);
+
+        return $this->persistProcessor->process($data, $operation, $uriVariables, $context);
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/StateProcessor/Put/AdminUserProcessor.php]: Shouldn't it be only for PUT operations?
```suggestion
        Assert::isInstanceOf($data, AdminUserInterface::class);
        Assert::isInstanceOf($operation, Put::class);

        $this->passwordUpdater->updatePassword($data);

        return $this->persistProcessor->process($data, $operation, $uriVariables, $context);
```

File: src/Sylius/Bundle/ApiBundle/StateProcessor/Put/AdminUserProcessor.php [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    public function process($data, Operation $operation, array $uriVariables = [], array $context = [])
+    {
+        Assert::isInstanceOf($data, AdminUserInterface::class);
+        Assert::isInstanceOf($operation, Put::class);
+
+        if ($operation instanceof DeleteOperationInterface) {
+            return $this->persistProcessor->process($data, $operation, $uriVariables, $context);
+        }
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/StateProcessor/Put/AdminUserProcessor.php]: These lines do not have sense after the above check:
```suggestion
```

**Review by GSadee — APPROVED** [review]

**Pull request merged**